osquery/tables/system/windows/windows_search.cpp

```diff
@@ -0,0 +1,244 @@
+// osquery/tables/system/windows/windows_search.cpp
+
+/**
+ * Copyright (c) 2014-present, The osquery authors
+ *
+ * This source code is licensed as defined by the LICENSE file found in the
+ * root directory of this source tree.
+ *
+ * SPDX-License-Identifier: (Apache-2.0 OR GPL-2.0-only)
+ */
+
+#include <atldbcli.h>
+#include <codecvt>
+#include <comutil.h>
+#include <osquery/core/core.h>
+#include <osquery/core/tables.h>
+#include <osquery/logger/logger.h>
+#include <osquery/utils/conversions/windows/strings.h>
+#include <osquery/utils/conversions/windows/windows_time.h>
+#include <sstream>
+#include <strsafe.h>
+#include <vector>
+#include <windows.h>
+
+#pragma comment(lib, "comsuppw.lib")
+
+namespace osquery {
+namespace tables {
+
+LONGLONG dateToUnixTime(const DATE date) {
+  SYSTEMTIME st;
+  FILETIME ft;
+
+  VariantTimeToSystemTime(date, &st);
+  SystemTimeToFileTime(&st, &ft);
+  return filetimeToUnixtime(ft);
+}
+
+// This helper function can print some propvariants and handles BSTR vectors
+void writePropVariant(REFPROPVARIANT variant, std::wstringstream& wss) {
+  if (variant.vt == (VT_ARRAY | VT_BSTR) && variant.parray->cDims == 1) {
```

Comment:
could be good to add a sanity check on the incoming type before using it. These variant types could be a bit scary sometimes

Reply:
could you expand on this a bit?

I thought that the default case at the bottom of the switch statement handle an unknown type?

Reply:
I asked chatgpt in an attempt to understand and it put this out

```cpp
  if ((variant.vt & VT_TYPEMASK) != VT_LPWSTR &&
      (variant.vt & VT_TYPEMASK) != VT_BSTR &&
      (variant.vt & VT_TYPEMASK) != VT_I1 &&
      (variant.vt & VT_TYPEMASK) != VT_UI2 &&
      (variant.vt & VT_TYPEMASK) != VT_I2 &&
      (variant.vt & VT_TYPEMASK) != VT_UI4 &&
      (variant.vt & VT_TYPEMASK) != VT_I4 &&
      (variant.vt & VT_TYPEMASK) != VT_UI8 &&
      (variant.vt & VT_TYPEMASK) != VT_I8 &&
      (variant.vt & VT_TYPEMASK) != VT_DATE &&
      (variant.vt & VT_TYPEMASK) != (VT_ARRAY | VT_BSTR)) {
    LOG(ERROR) << "Unhandled variant type " << variant.vt;
    return;
  }
```

did you have something like this in mind?

Reply:
I was thinking more about adding sanity checks to ensure that pointers are valid before dereferencing them. 

`variant.parray` should be checked for example before using it

Reply:
like a null check?
```cpp
if (variant == NULL) {
  // log error
}
```
sorry for all the questions, I'm coming from a GO background and still learning C++ best practices.

Reply:
sorry for the late response! Yes, exactly. The idea would be to check the pointers on the variant type before using them

Reply:
I tried this but variant does not have a `==` operator

Reply:
Thanks for looking into this! I was referring to checking the VARIANT's pointer member. In this specific case, I'll just add a check for NULL on `parray`

`if (variant.vt == (VT_ARRAY | VT_BSTR) && variant.parray && variant.parray->cDims == 1) {`

Reply:
thx! updated